lib/msprime.c

```diff
@@ -1763,9 +1759,365 @@ msp_insert_sample(msp_t *self, node_id_t sample, population_id_t population)
+                        ret = MSP_ERR_NO_MEMORY;
+                        goto out;
+                    }
+                    tail->next = seg;
+                    root_segments_tail[root] = seg;
+                }
+            }
+        }
+    }
+out:
+    return ret;
+}
+#endif
+
+static inline int
+msp_allocate_root_segments(msp_t *self, sparse_tree_t *tree,
+        uint32_t left, uint32_t right,
+        segment_t * restrict *root_segments_head,
+        segment_t * restrict *root_segments_tail)
+{
+    int ret = 0;
+    table_collection_t *tables = self->from_ts->tables;
+    const population_id_t *node_population = tables->nodes->population;
+    population_id_t population;
+    node_id_t root;
+    segment_t *seg, *tail;
+
+    for (root = tree->left_root; root != MSP_NULL_NODE; root = tree->right_sib[root]) {
+        population = node_population[root];
+        /* Reference integrity has alreay been checked, but the null population
+         * is still possibile */
+        if (population == MSP_NULL_POPULATION) {
+            ret = MSP_ERR_POPULATION_OUT_OF_BOUNDS;
+            goto out;
+        }
+        /* printf("Inserting seg %d-%d node=%d\n", left,right, root); */
+        if (root_segments_head[root] == NULL) {
+            seg = msp_alloc_segment(self, left, right, root, population,
+                    NULL, NULL);
+            if (seg == NULL) {
+                ret = MSP_ERR_NO_MEMORY;
+                goto out;
+            }
+            root_segments_head[root] = seg;
+            root_segments_tail[root] = seg;
+        } else {
+            tail = root_segments_tail[root];
+            if (tail->right == left) {
+                tail->right = right;
+            } else {
+                seg = msp_alloc_segment(self, left, right, root, population,
+                        tail, NULL);
+                if (seg == NULL) {
+                    ret = MSP_ERR_NO_MEMORY;
+                    goto out;
+                }
+                tail->next = seg;
+                root_segments_tail[root] = seg;
+            }
+        }
+    }
+out:
+    return ret;
+}
+
 static int
-msp_init_from_ts(msp_t *self)
+msp_reset_from_ts(msp_t *self)
 {
+    int ret = 0;
+    int t_iter;
+    sparse_tree_t t;
+    node_id_t root;
+    segment_t *seg;
+    uint32_t left, right, num_roots, overlap, last_overlap;
+    size_t num_nodes = self->tables.nodes->num_rows;
+    segment_t **root_segments_head = calloc(num_nodes, sizeof(*root_segments_head));
+    segment_t **root_segments_tail = calloc(num_nodes, sizeof(*root_segments_tail));
+
+    ret = sparse_tree_alloc(&t, self->from_ts, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    if (root_segments_head == NULL || root_segments_tail == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    /* Reset the tables to their correct position for replication */
+    ret = table_collection_reset_position(&self->tables, &self->from_position);
```

Comment:
This is a bit confusing. As far as I can tell, the tables can't have been modified since this position was set over [here](https://github.com/tskit-dev/msprime/pull/581/files#diff-b1775af7fb1c1f04ebe0aa2492b2953bR2141). Is this just paranoia?  Or could the tables have been actually modified?

Reply:
I think the confusion is because of the commented out code I've left lying around. But it's being reset at the start of every replicate back to the where the input tree sequence started.

Reply:
ah, re-use across replicates .I see. that's what "for replication" means. =)

Reply:
Yep -- there's a lot of effort in there to make replication as efficient as possible.